tests/vi_test.py

```diff
@@ -0,0 +1,281 @@
+    prior_mean = zeros(num_dim)
+    prior_cov = eye(num_dim)
+
+    x_o = zeros((1, num_dim))
+    target_distribution = true_posterior_linear_gaussian_mvn_prior(
+        x_o[0], likelihood_shift, likelihood_cov, prior_mean, prior_cov
+    )
+    target_samples = target_distribution.sample((num_samples,))
+
+    class TractablePotential(BasePotential):
+        def __call__(self, theta, **kwargs):
+            return target_distribution.log_prob(
+                torch.as_tensor(theta, dtype=torch.float32)
+            )
+
+        def allow_iid_x(self) -> bool:
+            return True
+
+    potential_fn = TractablePotential(prior=MultivariateNormal(prior_mean, prior_cov))
+    theta_transform = torch_tf.identity_transform
+
+    mu = zeros(num_dim, requires_grad=True)
+    scale = ones(num_dim, requires_grad=True)
+    q = torch.distributions.Independent(torch.distributions.Normal(mu, scale), 1)
+
+    posterior = VIPosterior(
+        potential_fn=potential_fn,
+        theta_transform=theta_transform,
+        q=q,
+        q_kwargs={"parameters": [mu, scale]},
+        vi_method="rKL",
+    )
+    posterior.set_default_x(x_o)
+    posterior.train(eps=1e-6)
+    samples = posterior.sample((num_samples,))
+    samples = torch.as_tensor(samples, dtype=torch.float32)
+
+    check_c2st(samples, target_samples, alg="slice_np")
+
+
+@pytest.mark.parametrize("q", ("maf", "nsf", "gaussian_diag", "gaussian", "mcf", "scf"))
+def test_deepcopy_support(q: str, set_seed):
+    """Tests if the variational does support deepcopy.
+
+    Args:
+        q: Different variational posteriors.
+    """
+
+    num_dim = 2
+
+    class FakePotential(BasePotential):
+        def __call__(self, theta, **kwargs):
+            return torch.ones_like(torch.as_tensor(theta, dtype=torch.float32))
+
+        def allow_iid_x(self) -> bool:
+            return True
+
+    potential_fn = FakePotential(prior=MultivariateNormal(zeros(num_dim), eye(num_dim)))
+    theta_transform = torch_tf.identity_transform
+
+    posterior = VIPosterior(
+        potential_fn=potential_fn,
+        theta_transform=theta_transform,
+        q=q,
+    )
+    posterior_copy = deepcopy(posterior)
+    posterior.set_default_x(torch.tensor(np.zeros((num_dim,)).astype(np.float32)))
+    assert posterior._x != posterior_copy._x, "Mhh, something with the copy is strange"
+    posterior_copy = deepcopy(posterior)
+    assert (
+        posterior._x == posterior_copy._x
+    ).all(), "Mhh, something with the copy is strange"
+
+    # Produces nonleaf tensors in the cache... -> Can lead to failure of deepcopy.
+    posterior.q.rsample()
+    posterior_copy = deepcopy(posterior)
+
+
+@pytest.mark.slow
+@pytest.mark.gpu
+@pytest.mark.parametrize("num_dim", (1, 2))
+@pytest.mark.parametrize("q", ("maf", "nsf", "gaussian_diag", "gaussian", "mcf", "scf"))
+@pytest.mark.parametrize("vi_method", ("rKL", "fKL", "IW", "alpha"))
+@pytest.mark.parametrize("sampling_method", ("naive", "sir"))
```

Comment:
How long do the additional tests take in total? I.e. all tests that are added in this PR

Reply:
I guess on my machine around 15 min.

Reply:
perfect!